irohad/ametsuchi/impl/tx_presence_cache_impl.hpp

```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_TX_PRESENCE_CACHE_IMPL_HPP
+#define IROHA_TX_PRESENCE_CACHE_IMPL_HPP
+
+#include "ametsuchi/storage.hpp"
+#include "ametsuchi/tx_presence_cache.hpp"
+#include "cache/cache.hpp"
+
+namespace iroha {
+  namespace ametsuchi {
+
+    class TxPresenceCacheImpl : public TxPresenceCache {
+     public:
+      explicit TxPresenceCacheImpl(std::shared_ptr<Storage> storage);
+      TxCacheStatusType check(
+          const shared_model::crypto::Hash &hash) const override;
+
+      BatchStatusCollectionType check(
+          const shared_model::interface::TransactionBatch &batch)
+          const override;
+
+     private:
+      std::shared_ptr<Storage> storage_;
+      mutable cache::Cache<shared_model::crypto::Hash,
```

Comment:
And, by the way, if you separate the `check` function into two parts, as I proposed earlier, `mutable` will no longer be needed